### PR TITLE
Evaluation

### PR DESCRIPTION
**Pull request by rpautrat**
Detector evaluation on the validation set of Coco and on HPatches.

### CODE REVIEW

File: superpoint/generate_coco_patches.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    print("Generating patches of Coco val...")
+    sess = tf.InteractiveSession()
+    for num, path in enumerate(image_paths):
+        new_path = Path(output_dir, str(num))
+        if not new_path.exists():
+            os.makedirs(new_path)
+
+        # Read the image
+        image = tf.read_file(str(path))
```
Comment (Skydes) [inline superpoint/generate_coco_patches.py]: Are you really creating new ops at every loop iteration? Shouldn't they be instantiated before hand? I don't think that even `tf.InteractiveSession` allows this.

File: superpoint/export_detections_repeatability.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        pbar = tqdm(total=config['eval_iter'] if config['eval_iter'] > 0 else None)
+        i = 0
+        while True:
+            try:
+                data = next(test_set)
+            except dataset.end_set:
+                break
+            data1 = {'image': data['image']}
+            data2 = {'image': data['warped_image']}
+            prob1 = net.predict(data1, keys='*')['prob_nms']
```
Comment (Skydes) [inline superpoint/export_detections_repeatability.py]: Consider writing 
```python
prob1 = net.predict(data1, keys='prob_nms')
```
instead, that should be faster.

File: superpoint/datasets/patches_dataset.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    def _init_dataset(self, **config):
+        dataset_folder = 'COCO/patches' if config['dataset'] == 'coco' else 'HPatches'
+        base_path = Path(DATA_PATH, dataset_folder)
+        folder_paths = [x for x in base_path.iterdir() if x.is_dir()]
+        image_paths = []
+        warped_image_paths = []
+        homographies = []
+        index_initial = len(str(base_path)) + 1
+        for path in folder_paths:
+            if config['alteration'] == 'i' and str(path)[index_initial] != 'i':
```
Comment (Skydes) [inline superpoint/datasets/patches_dataset.py]: `str(path)[index_initial]` could be simply replaced by `path.stem[0]`

File: superpoint/evaluations/detector_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+
+def compute_repeatability(exper_name, prob_thresh=0.5, distance_thresh=3):
+    """
+    Compute the repeatability. The experiment must contain in its output the prediction
+    on 2 images, an original image and a warped version of it, plus the homography
+    linking the 2 images.
+    """
+    def warp_keypoints(keypoints, H):
+        warped_col0 = np.add(np.sum(np.multiply(keypoints, H[0, :2]), axis=1), H[0, 2])
```
Comment (Skydes) [inline superpoint/evaluations/detector_evaluation.py]: Would be faster to convert to homogeneous coordinates, matrix multiply, and convert back to heterogeneous (see TF equivalent [here](https://github.com/rpautrat/SuperPoint/blob/master/superpoint/models/utils.py#L146)).

File: superpoint/evaluations/detector_evaluation.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+        # Warp the original keypoints with the true homography
+        true_warped_keypoints = warp_keypoints(keypoints[:, [1, 0]], H)
+        true_warped_keypoints[:, [0, 1]] = true_warped_keypoints[:, [1, 0]]
+        true_warped_keypoints = filter_keypoints(true_warped_keypoints, shape)
+
+        # Compute the repeatability
+        N1 = true_warped_keypoints.shape[0]
+        N2 = warped_keypoints.shape[0]
+        tile1 = np.tile(np.expand_dims(true_warped_keypoints, 1), (1, N2, 1))
```
Comment (Skydes) [inline superpoint/evaluations/detector_evaluation.py]: Tiling is unnecessary thanks to shape broadcasting (tensors still must have the rank though).

File: superpoint/evaluations/detector_evaluation.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        if N2 != 0:
+            min1 = np.min(norm, axis=1)
+            count1 = np.sum(min1 <= distance_thresh)
+        if N1 != 0:
+            min2 = np.min(norm, axis=0)
+            count2 = np.sum(min2 <= distance_thresh)
+        if N1 + N2 > 0:
+            repeatability.append((count1 + count2) / (N1 + N2))
+        else:
+            repeatability.append(1)
```
Comment (Skydes) [inline superpoint/evaluations/detector_evaluation.py]: I don't think that pairs with no common points should be taken into account at all, consider not appending their result.

**Comment by rpautrat** [comment]
Thanks, I have added all your suggestions. Shall we merge now?

**Comment by Skydes** [comment]
Looks good

**Pull request merged**